### PR TITLE
Make domains 8x larger (2x in each dimension), with origin at center of buildable space

### PR DESCRIPTION
**Pull request by ZappoMan**
This makes the center of the domain the origin, and makes the buildable space of the domain go from -16k,-16k,-16k to +16k,+16k,+16k

What is included:
* domain is now 32k in scale
* origin is at 0,0,0
* nothing moves

### CODE REVIEW

File: libraries/octree/src/OctreeElement.cpp [hunk 1]
```diff
@@ -191,6 +191,7 @@ void OctreeElement::calculateAACube() {
     // this tells you the "size" of the voxel
     float voxelScale = (float)TREE_SCALE / powf(2.0f, numberOfThreeBitSectionsInCode(getOctalCode()));
     corner *= (float)TREE_SCALE;
+    corner -= (float)HALF_TREE_SCALE;
```
Comment (ZappoMan) [inline libraries/octree/src/OctreeElement.cpp]: this is where all the magic happens

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4fe8f49d4ccd548641cf9c7a3fca61b254fe7378 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 85603dab29a622d208b4030858ed1e8df36eb4f5 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

**Comment by jherico** [comment]
What's the purpose of this change?

**Comment by ZappoMan** [comment]
@jherico - the purpose is to make @samcake and @PhilipRosedale happy

**Comment by PhilipRosedale** [comment]
The reason is that for a person building or exploring a domain, of course the only 'special' location one can imagine looking for is <0,0,0>.  But if you go there, you are actually at the edge of the domain, and 7/8th of the solid space around you is unbuildable.   

The correct design is that the origin represents the center of a buildable space.

File: libraries/entities/src/AddEntityOperator.cpp [hunk 2]
```diff
@@ -25,6 +25,8 @@ AddEntityOperator::AddEntityOperator(EntityTree* tree,
 {
     // caller must have verified existence of newEntity
     assert(_newEntity);
+    
+    // FIXME - how does this change for new origin???
     _newEntityBox = _newEntity->getMaximumAACube().clamp(0.0f, (float)TREE_SCALE);
```
Comment (AndrewMeadows) [inline libraries/entities/src/AddEntityOperator.cpp]: This would change to:

    _newEntityBox = _newEntity->getMaximumAACube().clamp((float)(-HALF_TREE_SCALE), (float)HALF_TREE_SCALE);

File: libraries/entities/src/EntitySimulation.cpp [hunk 3]
```diff
@@ -113,7 +113,8 @@ void EntitySimulation::sortEntitiesThatMoved() {
     // External changes to entity position/shape are expected to be sorted outside of the EntitySimulation.
     PerformanceTimer perfTimer("sortingEntities");
     MovingEntitiesOperator moveOperator(_entityTree);
-    AACube domainBounds(glm::vec3(0.0f,0.0f,0.0f), (float)TREE_SCALE);
+    AACube domainBounds(glm::vec3((float)-HALF_TREE_SCALE, (float)-HALF_TREE_SCALE, (float)-HALF_TREE_SCALE), 
+                            (float)TREE_SCALE);
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntitySimulation.cpp]: You can use the glm::vec3 constructor that takes only one argument for the corner argument.

File: libraries/octree/src/OctreeElement.cpp (hunk 1 above)
Comment (AndrewMeadows) [inline libraries/octree/src/OctreeElement.cpp]: This looks incorrect.  **corner** was being scaled by TREE_SCALE but is now being decremented by HALF_TREE_SCALE with no scaling.  I think the line should be:

    corner = (float)TREE_SCALE * corner - glm::vec3((float)HALF_TREE_SCALE);

or alternatively:

    corner = (float)TREE_SCALE * (corner - glm::vec3(0.5f));

File: libraries/octree/src/OctreeConstants.h [hunk 4]
```diff
@@ -17,7 +17,8 @@
 
 const quint64 CHANGE_FUDGE = 1000 * 200; // useconds of fudge in determining if we want to resend changed voxels
 
-const int   TREE_SCALE = 16384; // ~10 miles.. This is the number of meters of the 0.0 to 1.0 voxel universe
+const int TREE_SCALE = 16384; // ~10 miles.. This is the number of meters of the 0.0 to 1.0 voxel universe
+const int HALF_TREE_SCALE = TREE_SCALE / 2;
```
Comment (AndrewMeadows) [inline libraries/octree/src/OctreeConstants.h]: The vast majority of our code uses **TREE_SCALE** for floating point math rather than integer math.  Perhaps these should be of type **float**?

File: interface/src/Util.cpp [hunk 5]
```diff
@@ -43,6 +43,8 @@ void renderWorldBox(gpu::Batch& batch) {
 
     auto transform = Transform{};
     batch.setModelTransform(transform);
+    
+    // FIXME - new origin tweaks need to be done to this
     geometryCache->renderLine(batch, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(TREE_SCALE, 0.0f, 0.0f), red);
     geometryCache->renderLine(batch, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, TREE_SCALE, 0.0f), green);
     geometryCache->renderLine(batch, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 0.0f, TREE_SCALE), blue);
```
Comment (AndrewMeadows) [inline interface/src/Util.cpp]: Yes, the endpoints of these lines need only be HALF_TREE_SCALE out.  Also it might be nice to have some negative lines drawn, perhaps with faded colors, dotted lines, or something to indicate their nature.

File: libraries/entities/src/EntityItem.cpp [hunk 6]
```diff
@@ -656,7 +660,7 @@ int EntityItem::readEntityDataFromBuffer(const unsigned char* data, int bytesLef
     } else {
         // legacy order of packing here
         // TODO: purge this logic in a few months from now (2015.07)
-        READ_ENTITY_PROPERTY(PROP_POSITION, glm::vec3, updatePosition);
+        READ_ENTITY_PROPERTY(PROP_POSITION, glm::vec3, updatePositionOldOrigin);
         READ_ENTITY_PROPERTY(PROP_DIMENSIONS, glm::vec3, updateDimensions);
         READ_ENTITY_PROPERTY(PROP_ROTATION, glm::quat, updateRotation);
         READ_ENTITY_PROPERTY(PROP_DENSITY, float, updateDensity);
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityItem.cpp]: This is the **else** block of the **if (args.bitstreamVersion >= VERSION_ENTITIES_HAVE_SIMULATION_OWNER_AND_ACTIONS_OVER_WIRE)** logic which was from a few (or several?) months ago.  Do we have a process or statute of limitations for culling crufty version codepaths?

Comment (ZappoMan) [inline libraries/entities/src/EntityItem.cpp]: the comment above says...

```
        // legacy order of packing here
        // TODO: purge this logic in a few months from now (2015.07)
```

So the answer is... around october we can purge.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5fdd7304aadc3148df9522f77f0fba692a2b7e16 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

File: libraries/octree/src/OctreeConstants.h (hunk 4 above)
Comment (ZappoMan) [inline libraries/octree/src/OctreeConstants.h]: I considered that change, but decided it was out of scope for this change request.

File: libraries/octree/src/OctreeElement.cpp (hunk 1 above)
Comment (ZappoMan) [inline libraries/octree/src/OctreeElement.cpp]: I don't understand your comment... the three code examples are identical in their results.

Comment (ZappoMan) [inline libraries/octree/src/OctreeElement.cpp]: corner is a value from 0 to 1.0... we scale that value by TREE_SCALE, which gives us a value of 0 to TREE_SCALE, which we then want to be re-aligned from -HALF_TREE_SCALE to HALF_TREE_SCALE, so we subtract HALF_TREE_SCALE, the result is that a corner value of 0 becomes -HALF_TREE_SCALE and a corner value of 0.5 becomes 0 and 1 becomes HALF_TREE_SCALE.

Comment (AndrewMeadows) [inline libraries/octree/src/OctreeElement.cpp]: Oh right sorry.  Somehow I got it in my mind that the green line was replacing the one above it.  Bad parsing of diff.

File: interface/src/Util.cpp (hunk 5 above)
Comment (ZappoMan) [inline interface/src/Util.cpp]: I changed the comment to include suggestion that maybe we'd want to draw something for negative build-able space, but that change was out of scope for this PR

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7c6a9b131d581644400879658d280435a11cfb22 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 44fcf71f075bfc3136c0e2c99f6beba523555227 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 65864b1876e4f1843c853f3e5ae28bbb5ce3bb48 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

File: examples/edit.js [hunk 7]
```diff
@@ -1019,9 +1019,13 @@ function getPositionToCreateEntity() {
...
     var cameraPosition = Camera.position;
+    
+    var HALF_TREE_SCALE = 16384;
 
-    var cameraOutOfBounds = cameraPosition.x < 0 || cameraPosition.y < 0 || cameraPosition.z < 0;
-    var placementOutOfBounds = placementPosition.x < 0 || placementPosition.y < 0 || placementPosition.z < 0;
+    var cameraOutOfBounds = cameraPosition.x < -HALF_TREE_SCALE || cameraPosition.y < -HALF_TREE_SCALE || 
+                            cameraPosition.z < -HALF_TREE_SCALE;
+    var placementOutOfBounds = placementPosition.x < -HALF_TREE_SCALE || placementPosition.y < -HALF_TREE_SCALE || 
+                            placementPosition.z < -HALF_TREE_SCALE;
```
Comment (AndrewMeadows) [inline examples/edit.js]: edit.js "out of bounds" is only checking against the negative sides of the domain cube, not the positive -- but this was true before this PR.  Doesn't block this PR I suppose but we should probably fix it someday.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 014bf661fa92de5c1c9126783a8d3cb84ed654fd available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

File: examples/edit.js (hunk 7 above)
Comment (ZappoMan) [inline examples/edit.js]: good point I'll add a bug in asana for @huffman - since he's Mr. Edit.js (I'm sure he loves that!)

Comment (huffman) [inline examples/edit.js]: So much!

But yes, this should be fixed... I'll fix it once this is merged to avoid conflicts.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 57150421023898dcf4dd0d9b71d7497b29f90759 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

**Comment by AndrewMeadows** [comment]
For the record (since someone was asking me if this PR was ready to merge) this PR introduces a 100% reliable crash in the linux debug interface.  The callstack looks something like this:

    #3  0x00007fffef20ac32 in __GI___assert_fail (assertion=0x11846e0 "clipValue < std::numeric_limits<int16_t>::max()", file=0x11843d0 "libraries/shared/src/SharedUtil.cpp", 
    line=559, function=0x1184d00 <packClipValueToTwoByte(unsigned char*, float)::__PRETTY_FUNCTION__> "int packClipValueToTwoByte(unsigned char*, float)") at assert.c:101
    #4  0x0000000000cae83e in packClipValueToTwoByte (buffer=0x9edeedc "\340\f", clipValue=33288.1289) at libraries/shared/src/SharedUtil.cpp:559
    #5  0x0000000000cf553e in OctreeQuery::getBroadcastData (this=0x7fffffffbfc8, destinationBuffer=0x9edeedc "\340\f") at libraries/octree/src/OctreeQuery.cpp:37
    #6  0x0000000000b01870 in Application::__lambda90::operator() (__closure=0x7fffffff9870, node=...) at interface/src/Application.cpp:3118
    #7  0x0000000000b0fee3 in LimitedNodeList::eachNode<Application::queryOctree(NodeType_t, PacketType::Value, NodeToJurisdictionMap&)::__lambda90>(Application::__lambda90) (this=0x1a96430, functor=...)
    at libraries/networking/src/LimitedNodeList.h:173
    #8  0x0000000000b020a6 in Application::queryOctree (this=0x7fffffffaaf0, serverType=111 'o', packetType=PacketType::EntityQuery, jurisdictions=...)
    at interface/src/Application.cpp:3124
    #9  0x0000000000b00264 in Application::update (this=0x7fffffffaaf0, deltaTime=0.0157427955) at interface/src/Application.cpp:2858

In short: the "far clip" of the "view transform" is too large to be compressed into a 16-bit integer.

**Comment by ZappoMan** [comment]
yes - the ball is in my court to fix this crash before we can/should merge

**Comment by hifi-gustavo** [comment]
Build console output for commit: 92ae3e758c9a849c60e13132a18566ea062732e4 available at: https://gist.github.com/76cf06a9c97d36a555b5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5555.dmg

**Comment by PhilipRosedale** [comment]
is this ready now @ZappoMan ?

**Comment by AndrewMeadows** [comment]
Looks good.  It no longer crashes.  I tested editing entities into the negative regions and restarted the entity-servers -- the content appears to be saved and loaded correctly.
